cpp/ql/src/semmle/code/cpp/ir/implementation/unaliased_ssa/internal/SimpleSSA.qll

```diff
@@ -1,12 +1,46 @@
 import AliasAnalysis
 import cpp
 private import semmle.code.cpp.ir.implementation.raw.IR
+private import semmle.code.cpp.ir.internal.IntegerConstant as Ints
 private import semmle.code.cpp.ir.internal.OperandTag
 private import semmle.code.cpp.ir.internal.Overlap
 
+private class IntValue = Ints::IntValue;
+
+private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, Type type, IntValue bitOffset) {
+  resultPointsTo(instr.getResultAddressOperand().getDefinitionInstruction(), var, bitOffset) and
+  type = instr.getResultType()
+}
+
+private predicate hasOperandMemoryAccess(MemoryOperand operand, IRVariable var, Type type, IntValue bitOffset) {
+  resultPointsTo(operand.getAddressOperand().getDefinitionInstruction(), var, bitOffset) and
+  type = operand.getType()
+}
+
+/**
+ * Holds if the specified variable should be modeled in SSA form. For unaliased SSA, we only model a variable if its
+ * address never escapes and all reads and writes of that variable access the entire variable using the original type
+ * of the variable.
+ */
+private predicate isVariableModeled(IRVariable var) {
+  not variableAddressEscapes(var) and
+  // There's no need to check for the right size. An `IRVariable` never has an `UnknownType`, so the test for
```

Comment:
What happens if the AST has an `UnknownType` for a variable? Is that case excluded in IR generation?

Reply:
I'd consider that an extractor bug. I'd prefer to fix or workaround those bugs when discovered. The enum size workaround in this PR and the decayed/inferred type fix from my previous PR are both instances of similar bugs, and both were caught by the IR sanity queries.